packages/loopring_v3/contracts/amm/libamm/AmmTransactionReceiver.sol

```diff
@@ -94,19 +101,110 @@ library AmmTransactionReceiver
             signature.offset := add(signature.offset, 0x20)
         }
         if (txType == AmmData.PoolTxType.JOIN) {
+            ctx.approveAmmUpdates(true);
             S.processJoin(
                 ctx,
                 abi.decode(data, (AmmData.PoolJoin)),
                 signature
             );
+            ctx.approveAmmUpdates(false);
         } else if (txType == AmmData.PoolTxType.EXIT) {
+            ctx.approveAmmUpdates(true);
             S.processExit(
                 ctx,
                 abi.decode(data, (AmmData.PoolExit)),
                 signature
             );
+            ctx.approveAmmUpdates(false);
+        } else if (txType == AmmData.PoolTxType.SET_VIRTUAL_BALANCES) {
+            ctx.approveAmmUpdates(true);
+            ctx.vTokenBalancesL2 = ctx.settings.controller.getVirtualBalances(ctx.tokenBalancesL2, ctx.vTokenBalancesL2);
```

Comment:
The pool contracts allow the virtual balances to be updated with arbitrary values by the controller, so it just depends on what the controller does.

`AmplifiedAmmController.sol` currently only supports normal and amplified pools. For pools where we want to update the virtual balances differently either another controller needs to be used for the pool or `AmplifiedAmmController.sol` needs to be upgraded to allow additional functionality. 

The reason I did not allow arbitrary virtual balances updates in the current controller is because there is not a single good way to do that as it will need to depend on external data/mechanism to do so. Just allowing the virtual balances to be updated to whatever values would of course not be very secure for LPs.